### PR TITLE
Implement Sensitive data type support

### PR DESCRIPTION
**Pull request by ekohl**
The Sensitive data type is implemented by copying Variant. While this isn't pretty (duplication is never great), it gets the job done.

One of the biggest changes is that in order to use Sensitive an entry in Hiera to convert must exist. This should be implemented in modules using data in modules but it can be implemented in the installer's Hiera config as well.

This is based on [Puppet's documentation](https://puppet.com/docs/puppet/6/securing-sensitive-data.html)

### CODE REVIEW

**Comment by ekohl** [comment]
The pin rdoc commit was already merged to master.

**Comment by ekohl** [comment]
Marking as draft so I can first test it end to end. Will mark as ready when I've verified it.

**Comment by evgeni** [comment]
Can't we move the code to some `WrapperDataType` class, and then make `Variant` and `Sensitive` basically just:

```ruby
module Kafo
  module DataTypes
    class Sensitive < WrapperDataType
    end

    DataType.register_type('Sensitive', Sensitive)
  end
end
```

Other than that, this looks rather logical to me (with my oh so deep Kafo knowledge *cough*)

**Comment by evgeni** [comment]
this passed tests, fwiw:

```patch
diff --git lib/kafo/data_type.rb lib/kafo/data_type.rb
index 007d32f..818db9d 100644
--- lib/kafo/data_type.rb
+++ lib/kafo/data_type.rb
@@ -111,6 +111,7 @@ module Kafo
   end
 end
 
+require 'kafo/data_types/wrapper_data_type'
 require 'kafo/data_types/aliases'
 require 'kafo/data_types/any'
 require 'kafo/data_types/array'
diff --git lib/kafo/data_types/sensitive.rb lib/kafo/data_types/sensitive.rb
index 5fd6d8d..f928988 100644
--- lib/kafo/data_types/sensitive.rb
+++ lib/kafo/data_types/sensitive.rb
@@ -1,48 +1,6 @@
 module Kafo
   module DataTypes
-    class Sensitive < DataType
-      def initialize(*inner_types)
-        @inner_types = inner_types.map { |t| DataType.new_from_string(t) }
-      end
-
-      def condition_value(value)
-        type = find_type(value)
-        type ? type.condition_value(value) : super(value)
-      end
-
-      def dump_default(value)
-        type = find_type(value)
-        type ? type.dump_default(value) : super(value)
-      end
-
-      def multivalued?
-        @inner_types.any? { |t| t.multivalued? }
-      end
-
-      def to_s
-        @inner_types.join(' or ')
-      end
-
-      def typecast(value)
-        type = find_type(value)
-        type ? type.typecast(value) : value
-      end
-
-      def valid?(value, errors = [])
-        type = find_type(value)
-        if type
-          type.valid?(value, errors)
-        else
-          errors << "#{value} is not one of #{to_s}"
-          false
-        end
-      end
-
-      private
-
-      def find_type(value)
-        @inner_types.find { |t| t.valid?(t.typecast(value)) }
-      end
+    class Sensitive < WrapperDataType
     end
 
     DataType.register_type('Sensitive', Sensitive)
diff --git lib/kafo/data_types/variant.rb lib/kafo/data_types/variant.rb
index 4301a7a..a01e130 100644
--- lib/kafo/data_types/variant.rb
+++ lib/kafo/data_types/variant.rb
@@ -1,48 +1,6 @@
 module Kafo
   module DataTypes
-    class Variant < DataType
-      def initialize(*inner_types)
-        @inner_types = inner_types.map { |t| DataType.new_from_string(t) }
-      end
-
-      def condition_value(value)
-        type = find_type(value)
-        type ? type.condition_value(value) : super(value)
-      end
-
-      def dump_default(value)
-        type = find_type(value)
-        type ? type.dump_default(value) : super(value)
-      end
-
-      def multivalued?
-        @inner_types.any? { |t| t.multivalued? }
-      end
-
-      def to_s
-        @inner_types.join(' or ')
-      end
-
-      def typecast(value)
-        type = find_type(value)
-        type ? type.typecast(value) : value
-      end
-
-      def valid?(value, errors = [])
-        type = find_type(value)
-        if type
-          type.valid?(value, errors)
-        else
-          errors << "#{value} is not one of #{to_s}"
-          false
-        end
-      end
-
-      private
-
-      def find_type(value)
-        @inner_types.find { |t| t.valid?(t.typecast(value)) }
-      end
+    class Variant < WrapperDataType
     end
 
     DataType.register_type('Variant', Variant)
diff --git lib/kafo/data_types/wrapper_data_type.rb lib/kafo/data_types/wrapper_data_type.rb
new file mode 100644
index 0000000..3902299
--- /dev/null
+++ lib/kafo/data_types/wrapper_data_type.rb
@@ -0,0 +1,48 @@
+module Kafo
+  module DataTypes
+    class WrapperDataType < DataType
+      def initialize(*inner_types)
+        @inner_types = inner_types.map { |t| DataType.new_from_string(t) }
+      end
+
+      def condition_value(value)
+        type = find_type(value)
+        type ? type.condition_value(value) : super(value)
+      end
+
+      def dump_default(value)
+        type = find_type(value)
+        type ? type.dump_default(value) : super(value)
+      end
+
+      def multivalued?
+        @inner_types.any? { |t| t.multivalued? }
+      end
+
+      def to_s
+        @inner_types.join(' or ')
+      end
+
+      def typecast(value)
+        type = find_type(value)
+        type ? type.typecast(value) : value
+      end
+
+      def valid?(value, errors = [])
+        type = find_type(value)
+        if type
+          type.valid?(value, errors)
+        else
+          errors << "#{value} is not one of #{to_s}"
+          false
+        end
+      end
+
+      private
+
+      def find_type(value)
+        @inner_types.find { |t| t.valid?(t.typecast(value)) }
+      end
+    end
+  end
+end
```

**Comment by ekohl** [comment]
Thanks! I've applied that. I also tested this and it appears to work. I found some other issues in puppet-foreman that should be addressed, but this can be merged and released.

File: lib/kafo/data_types/sensitive.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require_relative 'wrapper_data_type'
+
+module Kafo
+  module DataTypes
+    class Sensitive < WrapperDataType
+    end
+
+    DataType.register_type('Sensitive', Sensitive)
```
[the remark this answers is not in the document]
Comment (ekohl) [reply lib/kafo/data_types/sensitive.rb:8]: Oh, that's an interesting thought. I like it.

[the remark this answers is not in the document]
Comment (evgeni) [reply lib/kafo/data_types/sensitive.rb:8]: Btw, what's more correct, wrappeR as it wraps, or wrappeD as it is wrapped?

[the remark this answers is not in the document]
Comment (ekohl) [reply lib/kafo/data_types/sensitive.rb:8]: I also thought about `ContainerDataType` but I'm not sure if that makes sense.

[the remark this answers is not in the document]
Comment (evgeni) [reply lib/kafo/data_types/sensitive.rb:8]: Naming things is hard

[the remark this answers is not in the document]
Comment (ekohl) [reply lib/kafo/data_types/sensitive.rb]: So how about this?

**Comment by evgeni** [comment]
This looks good to me, but I am hesitant to ACK as my kafo knowledge is rather limited

@ares, what do you think?

**Pull request merged**